packages/tree/src/Tree.tsx

```diff
@@ -0,0 +1,373 @@
+  expandAll = false,
+  parentId,
+  selectable = false
+}: TreeProps): JSX.Element | null => {
+  const [treeItems, setTreeItems] = useState(items);
+  const [selectedList, setSelectedList] = useState(selectedItems);
+  const [allExpanded, setAllExpanded] = useState(expandAll);
+
+  const areAllExpanded = (items: TreeItem[]) => {
+    let allExpanded = items.every((item) => item.isExpanded);
+    for (const item of items) {
+      if (item.children) {
+        allExpanded = allExpanded && areAllExpanded(item.children);
+      }
+    }
+    return allExpanded;
+  };
+
+  const areAllSelected = useCallback((items: TreeItem[]) => {
+    let allSelected = items.every((item) => item.isSelected);
+
+    for (const item of items) {
+      if (item.children) {
+        allSelected = allSelected && areAllSelected(item.children);
+      }
+    }
+    return allSelected;
+  }, []);
+
+  const [rootSelectText, setRootSelectText] = useState(SELECT_ALL);
+  const [rootExpandAllText, setRootExpandAllText] = useState(expandAll ? COLLAPSE_ALL : EXPAND_ALL);
+  const [isReady, setIsReady] = useState(false);
+
+  const walkTree = useCallback((item: TreeItem, fn: (item: TreeItem) => void) => {
+    fn(item);
+    if (item.children) {
+      for (const child of item.children) {
+        walkTree(child, fn);
+      }
+    }
+  }, []);
+
+  useEffect(() => {
+    if (treeItems) {
+      if (!isReady) {
+        for (const item of treeItems) {
+          item.isExpanded = item.isExpanded || expandAll || false;
+          item.isHidden = item.isHidden || false;
+          item.isDisabled = item.isDisabled || false;
+          item.isSelected = selectedItems.map((item) => item.id).includes(item.id) || item.isSelected || false;
+          item.areAllChildrenSelected = areAllChildrenSelected(item);
+        }
+      }
+      setIsReady(true);
+    }
+  }, [treeItems, expandAll, isReady, selectedItems]);
+
+  const handleItemSelected = useCallback(() => {
+    if (onItemsSelected) {
+      onItemsSelected(selectedList);
+    }
+  }, [selectedList, onItemsSelected]);
+
+  useEffect(() => {
+    if (areAllSelected(treeItems)) {
+      setRootSelectText(DESELECT_ALL);
+    } else {
+      setRootSelectText(SELECT_ALL);
+    }
+
+    for (const item of treeItems) {
+      item.areAllChildrenSelected = areAllChildrenSelected(item);
+    }
+
+    handleItemSelected();
+  }, [selectedList, treeItems, areAllSelected, handleItemSelected]);
+
+  useEffect(() => {
+    if (selectedItems.length > 0) {
+      setSelectedList(selectedItems);
+      for (const item of treeItems) {
+        walkTree(item, (item) => {
+          item.areAllChildrenSelected = areAllChildrenSelected(item);
+        });
+      }
+    }
+  }, [selectedItems, walkTree, treeItems]);
+
+  const filterItems = (items: TreeItem[], searchTerm: string) => {
+    const searchValue = searchTerm.toUpperCase();
+    const rootItems = items.filter((item) => item.name.toUpperCase().includes(searchValue));
```

Comment:
We have utilized some of this logic throughout RCM for a few years now and there hasn't been anything crazy that we have found. Obviously this might be different in the react world, and we don't have anything that super large in depth. We can definitively look into improving it in the future

Reply:
If y'all haven't seen any problems with it then that's good enough for me.